File: DSA_Level-2/HashMap And Heaps/Largest Subarray With Zero Sum/LargestSubarrayWithZeroSum_Solution.java
import java.util.*;

class Main {
	
	public static int solution(int[] arr) {
		
		Map<Integer,Integer> map=new HashMap<>();
		map.put(0,-1);

		int maxLength=0,sum=0;

		for(int i=0;i<arr.length;++i){
			sum+=arr[i];
			if(map.containsKey(sum)){
				maxLength=i-map.get(sum)>maxLength?i-map.get(sum):maxLength;
			}
			else{
				map.put(sum,i);
			}
		}


		return maxLength;

	}

	public static void main(String[] args) {
		Scanner scn = new Scanner(System.in);
		int[] arr = new int[scn.nextInt()];
		for (int i = 0; i < arr.length; i++) {
			arr[i] = scn.nextInt();
		}
		System.out.println(solution(arr));
	}

}

File: DSA_Level-2/HashMap And Heaps/Largest Subarray With Zero Sum/LargestSubarrayWithZeroSum_Solution.js

function largestSubarray(arr){
    
    let map=new Map();

    let sum=0;
    let maxLength=0;
    map.set(sum,-1);

    for(let i=0;i<arr.length;++i){
        
        sum+=arr[i];
        if(map.has(sum)){
            maxLength=Math.max(maxLength,i-map.get(sum));
        }
        else
            map.set(sum,i);

    }

    console.log(map);

    console.log(maxLength);

}

largestSubarray([15,-2,2,-8,1,7,10,23]);

